### PR TITLE
powerdns-admin: 0.3.0 -> 0.4.1

### PR DESCRIPTION
**Pull request by Flakebi**
###### Description of changes

Release notes: https://github.com/PowerDNS-Admin/PowerDNS-Admin/releases

- Fix with flask-migrate 4+
- Update to 0.4.1
- Improve the test to check that using the database works

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

ZHF: #230712

### CODE REVIEW

File: pkgs/development/python-modules/captcha/default.nix [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{ lib, fetchPypi, buildPythonPackage, nose, pillow }:
+
+buildPythonPackage rec {
+  pname = "captcha";
+  version = "0.4";
```
Comment (mweinelt) [inline pkgs/development/python-modules/captcha/default.nix:5]: Set `format = "setuptools"` explicitly.

File: pkgs/development/python-modules/captcha/default.nix [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+  src = fetchPypi {
+    inherit pname version;
+    hash = "sha256-KuXo2qxPFkm1ezQyi8xFumkbXHB/xvvdAW4hOuzpqLg=";
+  };
+
+  propagatedBuildInputs = [ pillow ];
+
+  nativeCheckInputs = [ nose ];
```
Comment (mweinelt) [inline pkgs/development/python-modules/captcha/default.nix:14]: Define an explicit `checkPhase`, so we're not relying on `python setup.py test`, which is deprecated.

https://github.com/lepture/captcha/blob/master/tox.ini#L9

**Review by mweinelt — CHANGES_REQUESTED** [review]
The styling comments apply to all the python packages. Also `pythonPackages` refers to `python2.pkgs`. You'll want `python310Packages` right now, so that the packages also get built by ofBorg.

**Comment by Flakebi** [comment]
Thanks for the references, I got tests working for all packages except for `flask-session-captcha`, where they fail with `RuntimeError: Working outside of application context`.

File: nixos/modules/services/web-apps/powerdns-admin.nix [hunk 3]
```diff
@@ -78,7 +78,7 @@ in
       environment.PYTHONPATH = pkgs.powerdns-admin.pythonPath;
       serviceConfig = {
         ExecStart = "${pkgs.powerdns-admin}/bin/powerdns-admin --pid /run/powerdns-admin/pid ${escapeShellArgs cfg.extraArgs}";
-        ExecStartPre = "${pkgs.coreutils}/bin/env FLASK_APP=${pkgs.powerdns-admin}/share/powerdnsadmin/__init__.py ${pkgs.python3Packages.flask}/bin/flask db upgrade -d ${pkgs.powerdns-admin}/share/migrations";
+        ExecStartPre = "${pkgs.coreutils}/bin/env FLASK_APP=${pkgs.powerdns-admin}/share/powerdnsadmin/__init__.py SESSION_TYPE= ${pkgs.python3Packages.flask}/bin/flask db upgrade -d ${pkgs.powerdns-admin}/share/migrations";
```
[the remark this answers is not in the document]
Comment (Flakebi) [reply nixos/modules/services/web-apps/powerdns-admin.nix:81]: These are only set for the Pre command, not for the main command.

File: nixos/modules/services/web-apps/powerdns-admin.nix [hunk 4]
```diff
@@ -78,7 +78,8 @@ in
       environment.PYTHONPATH = pkgs.powerdns-admin.pythonPath;
       serviceConfig = {
         ExecStart = "${pkgs.powerdns-admin}/bin/powerdns-admin --pid /run/powerdns-admin/pid ${escapeShellArgs cfg.extraArgs}";
-        ExecStartPre = "${pkgs.coreutils}/bin/env FLASK_APP=${pkgs.powerdns-admin}/share/powerdnsadmin/__init__.py ${pkgs.python3Packages.flask}/bin/flask db upgrade -d ${pkgs.powerdns-admin}/share/migrations";
+        # Set environment variables for starting through flask
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/powerdns-admin.nix:81]:
```suggestion
        # Set environment variables only for starting flask database upgrade
```

**Comment by SuperSandro2000** [comment]
We can undraft this, right?

**Comment by RaitoBezarius** [comment]
@zhaofengli please can you take a look?

**Comment by RaitoBezarius** [comment]
Result of `nixpkgs-review pr 227642` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>17 packages built:</summary>
  <ul>
    <li>powerdns-admin</li>
    <li>python310Packages.captcha</li>
    <li>python310Packages.captcha.dist</li>
    <li>python310Packages.flask-session-captcha</li>
    <li>python310Packages.flask-session-captcha.dist</li>
    <li>python310Packages.flask-sessionstore</li>
    <li>python310Packages.flask-sessionstore.dist</li>
    <li>python310Packages.wheezy-captcha</li>
    <li>python310Packages.wheezy-captcha.dist</li>
    <li>python311Packages.captcha</li>
    <li>python311Packages.captcha.dist</li>
    <li>python311Packages.flask-session-captcha</li>
    <li>python311Packages.flask-session-captcha.dist</li>
    <li>python311Packages.flask-sessionstore</li>
    <li>python311Packages.flask-sessionstore.dist</li>
    <li>python311Packages.wheezy-captcha</li>
    <li>python311Packages.wheezy-captcha.dist</li>
  </ul>
</details>

**Review by RaitoBezarius — APPROVED** [review]

**Comment by zhaofengli** [comment]
Sorry, totally forgot about this PR despite seeing the email a while ago :pensive: Haven't tested but the diff seems reasonable. Thanks for the work!

**Pull request merged**